packages/core/content-releases/server/src/controllers/release.ts

```diff
@@ -29,50 +28,110 @@ const releaseController = {
-      const query: GetContentTypeEntryReleases.Request['query'] =
-        await permissionsManager.sanitizeQuery(ctx.query);
+    const query = await permissionsManager.sanitizeQuery(ctx.query);
+    const { contentTypeUid, documentId, hasEntryAttached, locale } = query;
+    const entryId = await getEntryId({ contentTypeUid, documentId, locale }, { strapi });
 
-      const contentTypeUid = query.contentTypeUid;
-      const entryId = query.entryId;
-      // Parse the string value or fallback to a default
-      const hasEntryAttached: GetContentTypeEntryReleases.Request['query']['hasEntryAttached'] =
-        typeof query.hasEntryAttached === 'string' ? JSON.parse(query.hasEntryAttached) : false;
-
-      const data = hasEntryAttached
-        ? await releaseService.findManyWithContentTypeEntryAttached(contentTypeUid, entryId)
-        : await releaseService.findManyWithoutContentTypeEntryAttached(contentTypeUid, entryId);
-
-      ctx.body = { data };
-    } else {
-      const query: GetReleases.Request['query'] = await permissionsManager.sanitizeQuery(ctx.query);
-      const { results, pagination } = await releaseService.findPage(query);
+    if (!entryId) {
+      throw new errors.NotFoundError('Entry not found');
+    }
 
-      const data = results.map((release: ReleaseWithPopulatedActions) => {
-        const { actions, ...releaseData } = release;
+    const isEntryAttachedBoolean =
+      typeof hasEntryAttached === 'string' ? JSON.parse(hasEntryAttached) : false;
 
-        return {
-          ...releaseData,
+    if (isEntryAttachedBoolean) {
+      const releases = await releaseService.findMany({
+        where: {
           actions: {
-            meta: {
-              count: actions.count,
+            target_type: contentTypeUid,
+            target_id: entryId,
+          },
+          releasedAt: {
+            $null: true,
+          },
+        },
+        populate: {
+          actions: {
+            fields: ['type'],
+            where: {
+              target_type: contentTypeUid,
+              target_id: entryId,
             },
           },
-        };
+        },
       });
 
-      const pendingReleasesCount = await strapi.db.query(RELEASE_MODEL_UID).count({
+      ctx.body = { data: releases };
+    } else {
+      const relatedReleases = await releaseService.findMany({
         where: {
-          releasedAt: null,
+          releasedAt: {
+            $null: true,
+          },
+          actions: {
+            target_type: contentTypeUid,
+            target_id: entryId,
+          },
+        },
+      });
+
+      const releases = await releaseService.findMany({
+        where: {
+          $or: [
+            {
+              id: {
+                $notIn: relatedReleases.map((release: any) => release.id),
+              },
+            },
+            {
+              actions: null,
+            },
+          ],
+          releasedAt: {
+            $null: true,
+          },
```

Comment:
I believe these queries could be inside the service itself.

releases.findEntryReleases({id, populate...})
releases.findNotInEntryReleases({id, ...))

You could also reduce logic here if you do something like:
releases.findDocumentReleases({documentId, locale, populate...})
releases.findNotInDocumentReleases({documentId, locale, ...))

Reply:
I'm not 100% sure about this but we can discuss it when you're back

It's true that we reduce the controller complexity but we increase the service complexity and we need to maintain and look up for breaking changes in the service but not in the controller. What I mean is: Users can use the releases service in plugins or other places so in my opinion service's API should be as simple as possible and all these functions are just different way of using the findMany that already exists in the service